### PR TITLE
Prevent deletion of pages with children

### PR DESCRIPTION
**Pull request by luca-rath**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | -
| Related issues/PRs | #6104
| License | MIT
| Documentation PR | -

#### What's in this PR?

Prevent deletion of pages with children

### CODE REVIEW

File: src/Sulu/Bundle/SecurityBundle/AccessControl/AccessControlQueryEnhancer.php [hunk 1]
```diff
@@ -109,4 +110,57 @@ private function enhanceQueryWithAccessControl(
...
+     * @param array<string|int> $entityIds
+     *
+     * @return array<string|int>
+     */
+    public function findIdsWithGrantedPermissions(
+        ?UserInterface $user,
+        int $permission,
+        string $entityClass,
+        array $entityIds
+    ): array {
```
Comment (luca-rath) [inline src/Sulu/Bundle/SecurityBundle/AccessControl/AccessControlQueryEnhancer.php]: Not sure where we should move this method

Comment (luca-rath) [reply src/Sulu/Bundle/SecurityBundle/AccessControl/AccessControlQueryEnhancer.php]: Maybe the `AccessControlRepository`? as @nnatter mentioned in [this comment](https://github.com/sulu/sulu/pull/6123#discussion_r666194309).

File: src/Sulu/Bundle/PageBundle/EventListener/PageRemoveSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        foreach ($result->getRows() as $row) {
+            $descendantPages[] = [
+                'id' => $row->getValue('page.jcr:uuid'),
+                'depth' => \substr_count((string) $row->getPath(), '/'),
+                'resourceKey' => PageDocument::RESOURCE_KEY,
+            ];
+        }
+
+        $descendantPageIds = \array_column($descendantPages, 'id');
+        $descendantAuthorizedPageIds = $this->accessControlQueryEnhancer->findIdsWithGrantedPermissions(
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/PageBundle/EventListener/PageRemoveSubscriber.php]: this function in the enhancer is strange - you have mentioned it already - i would move it into a repository

File: src/Sulu/Bundle/PageBundle/EventListener/PageRemoveSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+            $grouped[$depth][] = $resource;
+        }
+
+        \krsort($grouped);
+
+        return \array_values($grouped);
+    }
+
+    public function handleRemove(RemoveEvent $event): void
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/PageBundle/EventListener/PageRemoveSubscriber.php]: this should only be `validateIfReadyToRemove`? Or something? `handleRemove` sounds like the actual remove

File: src/Sulu/Bundle/SecurityBundle/AccessControl/AccessControlQueryEnhancer.php (hunk 1 above)
Comment (wachterjohannes) [reply src/Sulu/Bundle/SecurityBundle/AccessControl/AccessControlQueryEnhancer.php]: do it :)

**Review by wachterjohannes — CHANGES_REQUESTED** [review]

**Comment by luca-rath** [comment]
Failing `PHP Lint` has nothing to do with this pull request

**Review by wachterjohannes — APPROVED** [review]

**Pull request merged**